cpp/src/arrow/dataset/file_base.cc

```diff
@@ -41,5 +44,54 @@ Status FileBasedDataFragment::GetTasks(std::shared_ptr<ScanContext> scan_context
   return format_->ScanFile(location_, scan_options_, scan_context, out);
 }
 
+DirectoryBasedDataSource::DirectoryBasedDataSource(const FileSource& location,
+                                                   std::vector<fs::FileStats> stats,
+                                                   std::shared_ptr<FileFormat> format,
+                                                   std::shared_ptr<ScanOptions> opts)
+    : location_(location),
+      stats_(std::move(stats)),
+      format_(std::move(format)),
+      scan_options_(std::move(opts)) {}
+
+Status DirectoryBasedDataSource::Make(const FileSource& location,
+                                      std::shared_ptr<FileFormat> format,
+                                      std::shared_ptr<ScanOptions> opts,
+                                      std::unique_ptr<DirectoryBasedDataSource>* out) {
+  std::vector<fs::FileStats> stats;
+  fs::Selector s;
+  s.base_dir = location.path();
+  s.recursive = true;
+  RETURN_NOT_OK(location.filesystem()->GetTargetStats(s, &stats));
+
+  auto new_end =
+      std::remove_if(stats.begin(), stats.end(), [&](const fs::FileStats& stats) {
+        return stats.type() != fs::FileType::File ||
+               !format->IsKnownExtension(stats.extension());
```

Comment:
IMHO, globbing is much better than extension based discovery. I'd say open a high priority followup ticket or implement it immediately.

Reply:
[picomatch](https://github.com/micromatch/picomatch) is MIT licensed, I could port it and add it to the filesystem API

Reply:
@pitrou could you comment?

Reply:
https://issues.apache.org/jira/browse/ARROW-6257